### DISCUSSION TITLE
New Binary from Vec or Slice

### DISCUSSION THREAD

**Discussion by stofel** [rusterlium/rustler#530] [Q&A]
Is it possible add function to create binaries from rust &[u8]  without memcpy via copy_from_slice()?

**Comment by filmor** [comment]
You can use resource binaries for this:
- https://www.erlang.org/doc/man/erl_nif.html#enif_make_resource_binary
- https://docs.rs/rustler/latest/rustler/resource/struct.ResourceArc.html#method.make_binary

**Comment by stofel** [comment]
Need example, how to do it. I spend two 2days without result..
```#[rustler::nif]
fn genbin<'a>(env: Env<'a>, bin: Binary<'a>) -> NifResult<Term<'a>> {
    let bb = bin.as_slice();
    let size = bb.len();
    let v = bb.to_vec();
    let (buf, term) = unsafe { new_binary(env, size) };
    let now = Instant::now();
    unsafe {
        std::ptr::copy_nonoverlapping(v.as_ptr(), buf, size);
    }
    let elapsed = now.elapsed();
    let newb = NewBinary {buf, term, size};
    println!("Full Elapsed: {:.2?}", elapsed);
    Ok(Binary::from(newb).to_term(env))
}   
```
code like this make unnecessary copy

**Reply by filmor** [reply]
Please read through the links I gave you and try to build something from that. You can't "just" use a random buffer of unknown lifetime with Erlang. You have to create a resource, which is essentially an opaque handle in Erlang to track a foreign object, and then attach a resource binary to it. To make this safe, the lifetime of the referenced slice must be less or equal the lifetime of the resource. The tests for this are here:
https://github.com/rusterlium/rustler/blob/f1aaa3b6d886f0685489c6ffaffb2142d9f7e861/rustler_tests/native/rustler_test/src/test_resource.rs#
